livekit-plugins/livekit-plugins-google/livekit/plugins/google/llm.py

```diff
@@ -0,0 +1,298 @@
+        )
+
+        contents = _build_vertex_context(chat_ctx, id(self))
+
+        cmp = model.generate_content_async(
+            contents=contents,
+            generation_config=generation_config,
+            tools=tools or None,
+            stream=True,
+        )
+
+        return LLMStream(
+            llm=self,
+            vai_stream=cmp,
+            chat_ctx=chat_ctx,
+            fnc_ctx=fnc_ctx,
+        )
+
+
+def _extract_system_instruction(chat_ctx: llm.ChatContext):
+    system_messages = [msg.content for msg in chat_ctx.messages if msg.role == "system"]
+    if system_messages:
+        return system_messages
+    return None
+
+
+def _build_tools(fnc_ctx: llm.FunctionContext) -> List[Tool]:
+    tools = [
+        Tool(
+            function_declarations=[
+                FunctionDeclaration(
+                    name=fnc_info.name,
+                    description=fnc_info.description,
+                    parameters=_build_parameters(fnc_info.arguments),
+                )
+            ]
+        )
+        for fnc_info in fnc_ctx.ai_functions.values()
+    ]
+    return tools
+
+
+def _build_parameters(arguments: Dict[str, llm.FunctionArgInfo]) -> Dict[str, Any]:
+    properties = {
+        arg_name: {
+            "type": JSON_SCHEMA_TYPE_MAP.get(arg_info.type, "string"),
+            "description": arg_info.description,
+            **({"enum": arg_info.choices} if arg_info.choices else {}),
+        }
+        for arg_name, arg_info in arguments.items()
+    }
+    required = [
+        arg_name
+        for arg_name, arg_info in arguments.items()
+        if arg_info.default is inspect.Parameter.empty
+    ]
+
+    parameters: Dict[str, Any] = {
+        "type": "object",
+        "properties": properties,
+    }
+    if required:
+        parameters["required"] = required
+
+    return parameters
+
+
+class LLMStream(llm.LLMStream):
+    def __init__(
+        self,
+        llm: LLM,
+        *,
+        vai_stream: Awaitable[AsyncIterator[GenerationResponse]],
+        chat_ctx: llm.ChatContext,
+        fnc_ctx: Optional[llm.FunctionContext],
+    ) -> None:
+        super().__init__(llm, chat_ctx=chat_ctx, fnc_ctx=fnc_ctx)
+        self._awaitable_vai_stream = vai_stream
+        self._vai_stream: Optional[AsyncIterator[GenerationResponse]] = None
+        self._function_calls_info: list[llm.FunctionCallInfo] = []  # type: ignore
+
+    async def _main_task(self) -> None:
+        if not self._vai_stream:
+            self._vai_stream = await self._awaitable_vai_stream
+        try:
+            async for response in self._vai_stream:
+                for idx, candidate in enumerate(response.candidates):
+                    chat_chunk = self._parse_candidate(candidate, idx)
+                    if chat_chunk:
+                        self._event_ch.send_nowait(chat_chunk)
+        except Exception as e:
+            logger.error("Error during main task: %s", e)
+            raise
```

Comment:
We have an utility to directly log errors inside a task. 
Use [@utils.log_exceptions](https://github.com/livekit/agents/blob/1a99aae891d6bd440da20f88a3f6e29eab9069cc/livekit-agents/livekit/agents/utils/log.py#L7)